_posts/2020-09-29-introducing-torch-for-r/introducing_torch_for_R.Rmd

```diff
@@ -0,0 +1,424 @@
+
+author:
+  - name: Sigrid Keydana
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+slug: keydanaintroducingtorchforR
+bibliography: bibliography.bib
+date: 09-29-2020
+categories:
+  - Packages/Releases
+  - Torch
+  - R
+output:
+  distill::distill_article:
+    self_contained: false
+preview: images/torch.png
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+Last January at *rstudio::conf*, in that distant past when conferences still used to take place at some physical location, my colleague Daniel gave a talk introducing new features and ongoing development in the `tensorflow` ecosystem. In the Q&A part, he was asked something unexpected: Were we going to build support for PyTorch? He hesitated; that was in fact the plan, and he had already played around with natively implementing `torch` tensors at a prior time, but he was not completely certain how well "it" would work.
+
+"It", that is an implementation which does not bind to Python Torch, meaning, we don't install the PyTorch wheel and import it via `reticulate`. Instead, we delegate to the underlying C++ library `libtorch` for tensor computations and automatic differentiation, while neural network features -- layers, activations, optimizers -- are implemented directly in R. Removing the intermediary has at least two benefits: For one, the leaner software stack means fewer possible problems in installation and fewer places to look when troubleshooting. Secondly, through its non-dependence on Python, `torch` does not require users to install and maintain a suitable Python environment. Depending on operating system and context, this can make an enormous difference: For example, in many organizations employees are not allowed to manipulate privileged software installations on their laptops.
+
+So why did Daniel hesitate, and, if I recall correctly, give a not-too-conclusive answer? On the one hand, it was not clear whether compilation against `libtorch` would, on some operating systems, pose severe difficulties. (It did, but difficulties turned out to be surmountable; at some point, my colleague and `torch`-coauthor Javier might want to tell you about it.)[^1] On the other, the sheer amount of work involved in re-implementing -- not all, but a big amount of -- PyTorch in R seemed intimidating. Today, there is still lots of work to be done (we'll pick up that thread at the end), but the main obstacles have been ovecome, and enough components are available that `torch` can be useful to the R community. Thus, without further ado, let's train a neural network.
+
+[^1]: Lest there be any misunderstandings -- coauthor not to yours truly, but to Daniel, creator and main author of the package.
```

Comment:
Not sure if I'll ever write a post in the RStudio blog about lantern, but I think we can give a couple pointers to those interested!

```suggestion
So why did Daniel hesitate, and, if I recall correctly, give a not-too-conclusive answer? On the one hand, it was not clear whether compilation against `libtorch` would, on some operating systems, pose severe difficulties. (It did, but difficulties turned out to be surmountable; [Javier](https://twitter.com/javierluraschi) wrapped `libtorch` into [lantern](https://github.com/mlverse/lantern), a C interface to `libtorch` which avoids cross-compiler issues between MinGW and Visual Studio in Windows) On the other, the sheer amount of work involved in re-implementing -- not all, but a big amount of -- PyTorch in R seemed intimidating. Today, there is still lots of work to be done (we'll pick up that thread at the end), but the main obstacles have been ovecome, and enough components are available that `torch` can be useful to the R community. Thus, without further ado, let's train a neural network.
```

Reply:
hm... ok.... I think the alternative is too detailed though ... I'll see if I can "compress" it and/or move it to a footnote...